### PR TITLE
Add support for GHC 9.4.1

### LINKED ISSUE #21: 9.3 HEAD: Module ‘GHC.Tc.Types’ does not export ‘TcPluginResult(..)’
I hope this is some trivial API change and it would be easy to make ghc-tcplugins-extra work with head? I need this to be able to repro a suspect GHC bug with HEAD, but this will probably also make ghc-tcplugins-extra ready for GHC 9.4.1.

```
Failed to build ghc-tcplugins-extra-0.4.2.
Build log (
/home/mikolaj/.cabal/logs/ghc-9.3.20220316/ghc-tcplugins-extra-0.4.2-03780bc76cabaa0f8a3b8fb908cbadb752b4bbb38f36e6794751c3d459ceb2d9.log
):
Configuring library for ghc-tcplugins-extra-0.4.2..
Preprocessing library for ghc-tcplugins-extra-0.4.2..
Building library for ghc-tcplugins-extra-0.4.2..
[1 of 1] Compiling GHC.TcPluginM.Extra ( src/GHC/TcPluginM/Extra.hs, dist/build/GHC/TcPluginM/Extra.o, dist/build/GHC/TcPluginM/Extra.dyn_o )

src/GHC/TcPluginM/Extra.hs:64:37: error:
    Module ‘GHC.Tc.Types’ does not export ‘TcPluginResult(..)’
   |
64 | import GHC.Tc.Types (TcPlugin (..), TcPluginResult (..))
   |                                     ^^^^^^^^^^^^^^^^^^^
```

Looking at `compiler/GHC/Tc/Types.hs`

```Haskell
        -- Constraint solver plugins
        TcPlugin(..),
        TcPluginSolveResult(TcPluginContradiction, TcPluginOk, ..),
        TcPluginRewriteResult(..),
        TcPluginSolver, TcPluginRewriter,
        TcPluginM(runTcPluginM), unsafeTcPluginTcM,
```

perhaps the API just got more precise?

### PR DESCRIPTION
**Pull request by christiaanb**
Fixes #21

### CODE REVIEW

File: package.dhall [hunk 1]
```diff
@@ -22,16 +22,28 @@ in  let ghc = { name = "ghc", mixin = [] : List Text }
...
+                  , "GhcApi.Predicate"
+                  , "GhcApi.GhcPlugins"
+                  , "Internal.Type"
+                  , "Internal.Constraint"
+                  , "Internal.Evidence"
+                  , "Internal.TracePlugin"
+                  , "Internal.Module"
+                  ]
+
             in    defs
```
Comment (philderbeast) [inline package.dhall:36]: I'm noticing more indentation after running `dhall format package.dhall` with `dhall-1.41.1`.

File: src-ghc-tree/Internal.hs [hunk 2]
```diff
@@ -1,13 +1,14 @@
-{-# LANGUAGE RecordWildCards #-}
-{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE CPP #-}
```
Comment (philderbeast) [inline src-ghc-tree/Internal.hs:1]: With ghc-9.4 dropping derived constraints could we split the source again and avoid CPP use?

**Review by Mikolaj — APPROVED** [review]
I can barely even tell which parts of these commits are auto-generated, but FWIW, looks good to me.

**Comment by Mikolaj** [comment]
@philderbeast: do you think you could review this, pretty please? Thank you!

**Review by philderbeast — APPROVED** [review]

**Comment by philderbeast** [comment]
> @philderbeast: do you think you could review this, pretty please? Thank you!

Oh sorry. I left some comments a while back but didn't check one of the radio buttons to finish up.

**Comment by Mikolaj** [comment]
Happy times! We are so close after 5 months of struggle!

**Comment by Mikolaj** [comment]
@christiaanb: a very gentle, if impolite, reminder. :)

File: src-ghc-tree-9.4/Internal.hs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+      tcPluginTrace ("tcPluginInit " ++ s) empty >> tcPluginInit
+
+    traceStop  z = tcPluginTrace ("tcPluginStop " ++ s) empty >> tcPluginStop z
+
+    traceSolve z ev given wanted = do
+      tcPluginTrace ("tcPluginSolve start " ++ s)
+                        (text "given   =" <+> ppr given
+                      $$ text "wanted  =" <+> ppr wanted)
+      r <- tcPluginSolve z ev given wanted
+      case r of
```
[the remark this answers is not in the document]
Comment (christiaanb) [reply src-ghc-tree-9.4/Internal.hs:89]: Good point. I've decided to retain them.

File: src-ghc-tree-9.4/Internal.hs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+import Data.Maybe (mapMaybe)
+
+import GhcApi.Constraint (Ct(..), CtEvidence(..), CtLoc)
+import GhcApi.GhcPlugins
+
+import Internal.Type (substType)
+import Internal.Constraint (newGiven, flatToCt, mkSubst, overEvidencePredType)
+import Internal.Evidence (evByFiat)
+
+{-# ANN module "HLint: ignore Use fewer imports" #-}
```
[the remark this answers is not in the document]
Comment (christiaanb) [reply src-ghc-tree-9.4/Internal.hs]: Yes

File: src-ghc-tree-9.4/Internal.hs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+-- | Create a new [W]anted constraint.
+newWanted  :: CtLoc -> PredType -> TcPluginM CtEvidence
+newWanted = TcPlugin.newWanted
```
[the remark this answers is not in the document]
Comment (christiaanb) [reply src-ghc-tree-9.4/Internal.hs]: Done

**Pull request merged**